chainer/functions/noise/dropconnect.py

```diff
@@ -0,0 +1,121 @@
+import numpy
+
+import chainer
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _as_mat(x):
+    if x.ndim == 2:
+        return x
+    return x.reshape(len(x), -1)
+
+
+class Dropconnect(function.Function):
+
+    """Linear unit regularized by dropconnect."""
+
+    def __init__(self, ratio, mask=None):
+        self.ratio = ratio
+        self.mask = mask
+
+    def check_type_forward(self, in_types):
+        n_in = in_types.size()
+        type_check.expect(2 <= n_in, n_in <= 3)
+        x_type, w_type = in_types[:2]
+
+        type_check.expect(
+            x_type.dtype.kind == 'f',
+            w_type.dtype.kind == 'f',
+            x_type.ndim >= 2,
+            w_type.ndim == 2,
+            type_check.prod(x_type.shape[1:]) == w_type.shape[1],
+        )
+        if n_in.eval() == 3:
+            b_type = in_types[2]
+            type_check.expect(
+                b_type.dtype == x_type.dtype,
+                b_type.ndim == 1,
+                b_type.shape[0] == w_type.shape[0],
+            )
+
+    def forward(self, inputs):
+        scale = inputs[1].dtype.type(1. / (1 - self.ratio))
+        xp = cuda.get_array_module(*inputs)
+        if xp == numpy:
+            flag = xp.random.rand(*inputs[1].shape) >= self.ratio
```

Comment:
In honest, I just referred to `Dropout`.

According to here, the purpose seems to be memory saving.
https://github.com/pfnet/chainer/pull/620

However, it seems that float16 can be good, and it may be also applicable to numpy.